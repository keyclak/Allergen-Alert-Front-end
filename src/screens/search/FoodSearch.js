import React, { useEffect, useState } from "react";
import {Image, View, StatusBar, StyleSheet, Text, TouchableOpacity, Button, TextInput, SafeAreaView, FlatList, Pressable, ActivityIndicator, RefreshControl } from "react-native";
import { StyleConstants, Styles, Colors } from '../../style';
import { useGetFoodSearch } from '../../hooks/api';
import { useDummy } from '../../hooks/api';
import ButtonTextInput from '../../components/ButtonTextInput';
import FloatingButton from "../../components/FloatingButton";
import { MaterialIcons, MaterialCommunityIcons } from "@expo/vector-icons";
import { ScrollView } from "react-native-gesture-handler";
import FormTextInput from "../../components/FormTextInput";

export default function FoodSearch({navigation}) {
    const [searchValue, setSearchValue] = useState("");
    const getFoodSearch = useGetFoodSearch(searchValue);

    const fetchData = async () => {
        getFoodSearch.background(); 
    };

    useEffect(() => console.log(getFoodSearch.response), [searchValue]);

    function renderHeader() {
        return (
            <View
                style={{
                    backgroundColor: Colors.Blue[5],
                    borderRadius: StyleConstants.Radius,
                    height: 40,
                    flexDirection: 'row',
                    alignItems: 'center'
                }}>
                <TextInput
                    placeholder="Search"
                    placeholderTextColor={Colors.Gray[4]}
                    onChangeText={setSearchValue}
                    onSubmitEditing={() => fetchData()}
                    style={{
                        color: Colors.Gray[1],
                        marginLeft: StyleConstants.Radius * 0.75,
                        fontSize: StyleConstants.FormItemTextSize,
                        flexGrow: 1
                    }}/>
                
                <Pressable style={{width: 40, alignItems: 'center', }} onPress={() => fetchData()}>
                    <MaterialIcons name="search" color={Colors.Background} size={24} />
                </Pressable>
            </View>
        );
    }

    useEffect(() => {
        navigation.setOptions({
            headerTitle: renderHeader
        });
    }, [searchValue]);

    const Item = ({ item }) => (
        <Pressable style={Styles.listItem} onPress={() => navigation.navigate('FoodPage', { foodId: item.id })}>
            <Text style={Styles.listItemText}>{item.name}</Text>
        </Pressable>
    );

    function openScanner() {
        navigation.push('Scanner');
    }

    function getEmptyComponent() {
        return (
            <View style={{ height: '100%', flexGrow: 1, alignItems: 'center', justifyContent: 'center'}}>     
                {searchValue
                    ? <Text style={{color: Colors.Gray[5]}}>No results found</Text>
                    : <Text style={{color: Colors.Gray[5]}}>Search for something, or open the barcode scanner</Text>
                }
            </View>
        );
    }


    return (
        <View style={Styles.container}>
            <FlatList
                refreshControl={<RefreshControl colors={[Colors.Blue[4]]} refreshing={getFoodSearch.loading} onRefresh={fetchData}/>}
                ListEmptyComponent={getEmptyComponent}
                data={getFoodSearch.response}
                keyExtractor={item => item.id}
                renderItem={Item} 
                extraData={getFoodSearch.response}
                contentContainerStyle={{flexGrow: 1}}
                style={{width: '100%', alignContent: 'center'}}/>

            <View style={{ width: '100%', height: '100%', position: 'absolute', alignItems: 'flex-end', justifyContent: 'flex-end'}}>
                <FloatingButton onPress={openScanner}>
                    <MaterialCommunityIcons name="barcode-scan" size={32} color="white"/>
                </FloatingButton>
            </View>
        </View>
    );
}