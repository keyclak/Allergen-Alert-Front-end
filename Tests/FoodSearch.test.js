import renderer from 'react-test-renderer';
import React from 'react';
import MockDate from 'mockdate'
import FoodSearch from '../src/screens/search/FoodSearch.js';

const FRAME_TIME = 10

const advanceOneFrame = () => {
  const now = Date.now();
  MockDate.set(new Date(now + FRAME_TIME));
  jest.advanceTimersByTime(FRAME_TIME);
}

const timeTravel = (msToAdvance = FRAME_TIME) => {
  const numberOfFramesToRun = msToAdvance / FRAME_TIME;
  let framesElapsed = 0;

  while (framesElapsed < numberOfFramesToRun) {
    advanceOneFrame();
    framesElapsed++;
  }
}
global.requestAnimationFrame = cb => {
  setTimeout(cb, FRAME_TIME);
}

export const setupTimeTravel = () => {
    MockDate.set(0);
    jest.useFakeTimers();
}

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('@expo/vector-icons', () => ({
    MaterialIcons: 'Icon',
    MaterialCommunityIcons: 'Icon'
}));

beforeEach(setupTimeTravel);

const navigation = {
    addListener: jest.fn(),
    setOptions: jest.fn()
}

describe("FoodSearchPageTests", () => {
    test("Renders correctly", async () => {
        const tree = renderer.create(
            <FoodSearch navigation={navigation} />
        ).toJSON(); 

        timeTravel(500);

        expect(tree).toMatchSnapshot();
    })
})