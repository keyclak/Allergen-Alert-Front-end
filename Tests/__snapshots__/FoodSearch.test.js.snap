// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodSearchPageTests Renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={null}
    disableVirtualization={false}
    extraData={null}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "#2AB7CA",
          ]
        }
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "alignContent": "center",
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "height": "100%",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#adb5bd",
            }
          }
        >
          Search for something, or open the barcode scanner
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "height": "100%",
        "justifyContent": "flex-end",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#2AB7CA",
            "borderRadius": 30,
            "elevation": 5,
            "height": 60,
            "justifyContent": "center",
            "width": 60,
          },
          Object {
            "bottom": 30,
            "right": 30,
          },
        ]
      }
    >
      <Icon
        color="white"
        name="barcode-scan"
        size={32}
      />
    </View>
  </View>
</View>
`;
